Implements ResolveServicesActor like suggested in #37

Stelzi79 opened the pull request:
As proposed in #37 this PR implements the capability to resolve things from some sort of container system by sending a ```IResolveMessage<T>``` to an Actor that is inherited from ```ResolveActor```.

I have implemented the ```ResolveServicesActor``` that resolves Services form the built in dependency injection system of .net Core. It is contained in one additional ```netcoreapp3.0``` library project.

~~The ```ResolveActors``` Type (a Type to inject more easier multiple IActorRefs into DI)~~ (**Edit:** *wasn't fitting very well in the overall concepts of adding stuff to the ```Startup.cs``` and have it in the composing nature of it!*)

A ```IResolveActor``` for example can be added as a service to DI by:

```csharp
ConfigureServices(IServiceCollection services)
{
	// Adds a ResolveServiceActor to the DI system
	services.AddResolveActor<ResolveServicesActor>();
}
```

This makes them ready for use.
```csharp
Configure(IApplicationBuilder app, /*...*/)
{
	// Initializes all ResolveActors
	app.UseResolveActors();
}
```

If it is wanted that a special actor supervises these actors instead of the ActorSystem it can be added as following inside an Actor:

```csharp
// Adds ResolveServiceActor wit an IServiceProvider container 
// as a child to this special supervising Actor
IActorRef resolver = ResolveServicesActor.Create(container, Context);
```

Services can be received by 

```csharp
var service = ResolveServicesActor.Instance.Ask(new RequestService(typeof(IService)));
```
*Note: I will probably implement an ```IActorRef.RequestService<IService>()``` extension method to make it a bit cleaner*

The ```ResolveActor``` base class follows the convention of ```ChannelManagerActor```
of a static ```Create()``` method and a ```public static IActorRev Instance``` property.

This system of resolving things can be extended to also handle for example retrieving configurations, or similar things that are not easy retrievable by the very uncoupled ActorSystem.

This PR is currently a WIP! It is currently missing these things:

- [x] ~~```IActorRef.RequestService<IService>()```~~ (Wrong Interface to extend) ```ActorBase.RequestService<IService>()``` extension method
- [ ] Unit Tests - Yes it compiles at least on my machine!
- [x] All actors resolve there services from this ResolveServiceActor

Stelzi79 committed:
first implementing of Interfaces and Baseclasses. Implemented ResolveServicesActor to resolve from DI system.

Stelzi79 committed:
fixed spelling

Stelzi79 committed:
remove unneeded extension method

csharpfritz commented:
Great innovation here that will help the architecture of this project.  Please remove the WIP label when you are completed

Stelzi79 committed:
Added Extension Methods to ActorBase (IActorRef is the wrong type to extend!)

Stelzi79 committed:
fixe NullReferenceException on Instance when
trying to access it before the ActorSystem is started fully.
This is fixed by sending a InitResolveActor to Become() ready for resolving.
Sending this message shortly after adding activates the AcotrSystem and every
Actor added after that can request Services in the constructor.

Stelzi79 commented:
```IActorRef``` is actually the wrong Interface to extend. It makes much more sense to extend ```ActorBase```. This is the abstract base class of a ```ReceiveActor``` and a ```UntypedActor``` and is therefore more usable.

An ```IService``` can now already in the constructor of an Actor resolved by:

```csharp
var service = this.RequestService<IService>();

var services = this.RequestManyServices<IService>();
```

These synchronous extension Methods simply wrap asynchronous methods because ```IActorRef.Ask<T>``` is per definition asynchronous and is a bit naughty in that it doesn't follow the Async post fix convention.

There is also the ```RequestServiceAsync<IService>()``` and ```RequestManyServicesAsync<IService>()``` extension methods exposed if needed.

Stelzi79 committed:
readded params args handling in ResolveActor

Stelzi79 committed:
some cleanup and additional docs

Stelzi79 committed:
removed ResolveActors (wasn't fitting very well) and changed how to add resolvers to DI

Stelzi79 committed:
some cleanup and documentations

Stelzi79 committed:
added a IApplicationBuilder.UseResolveActors() to initialize all IResolveActors

Stelzi79 committed:
added implementations to project files

Stelzi79 committed:
FollowerServiceActor now resolves all services through RequestService

Stelzi79 committed:
added BotConfiguration to be resolved by everybody

Stelzi79 committed:
entangled ChannelConfigurationActor

Stelzi79 committed:
did the last cleanups. All Services should now be resolved by ResolveServicesActor

Stelzi79 committed:
some cleanup

Stelzi79 commented:
I think I have now got all Actors that request services from DI. I couldn't test it fully because of missing authentication configuration.

There are in the files I changed a few cleanups done. I have a VS extension that respects any settings in .editorconfig and does a format code on save.

* sort usings
* remove unused using
* replace spaces with tabs
* generally put { on new line

In principle it should be ready for usage.

Stelzi79 committed:
changed .gitignore to exclude .vscode/settings.json

Stelzi79 committed:
updated startup.cs to use the changes after a rebase

Stelzi79 committed:
some cleanups

Stelzi79 committed:
--ammend

Stelzi79 committed:
deleted old FollowerServiceActor - you can need git to revive it if needed
